### ISSUE TITLE
Haiku LSTM vs Torch LSTM

### ISSUE THREAD

**Issue by pranavsubramani** [deepmind/dm-haiku#70]
I'm trying to figure out the differences between pytorch's LSTM and Haiku's LSTM. In Haiku, the LSTM expects the input to be a rank 1 or rank 2 tensor. However, in pytorch, the input is expected to be a rank 3 tensor. I was hoping to get some clarification on how one would expect to get the same behavior in Haiku's lstms compared to torch's lstms.

Some additional information -- if this is the case, it seems like the LSTM in Haiku is significantly slower than torch. It's very likely that I've made a mistake somewhere here. 

Also, would it be correct in saying that haiku's `static_unroll` + `lstm` is the same as pytorch's `lstm`? In which case, isn't Haiku's lstm technically an LSTMcell?

**Comment by superbobry** [comment]
Hi Pranav,

Yes, `hk.LSTM` is a cell or a [*core*](https://dm-haiku.readthedocs.io/en/latest/api.html#recurrent-modules) in Haiku lingo, i.e. a single-step computation which expects a tensor of shape `[B, D]` or `[D]` where `B` is the batch dimension. In order to apply a core to the whole sequence (and get PyTorch's LSTM) you need to use one of the unroll functions, whether `static_*` or `dynamic_*` doesn't really matter.

Could you post the performance differences you've observed here and include the code (if possible)? 

Disclaimer: I would expect Haiku to be slower than PyTorch at least on GPU, because PyTorch uses a fused [CuDNN-RNN](https://docs.nvidia.com/deeplearning/cudnn/developer-guide/index.html#features-of-rnn-functions) kernel for both forward and backward computations, whereas Haiku implements LSTM only in terms of XLA primitives.

**Comment by tomhennigan** [comment]
Wrt benchmarks you may find that microbenchmarks of the LSTM module alone may be slightly slower than other frameworks (for the reasons @superbobry points out), but you should find performance to be competitive when you benchmark larger portions of your code (e.g. a complete agent, loss function or training step).

As with all JAX benchmarks be sure to (1) double check you are [running JAX on your GPU](https://github.com/google/jax/#pip-installation) (2) use `jax.jit` around as much code as possible (e.g. around your whole apply function from Haiku) and (3) account for JAX's [async dispatch](https://jax.readthedocs.io/en/latest/async_dispatch.html) (usually this will result in benchmarks looking way faster than expected).
